src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -26,13 +26,17 @@ namespace OpenTelemetry.Metrics
     internal sealed class AggregatorStore
     {
         private static readonly ObjectArrayEqualityComparer ObjectArrayComparer = new ObjectArrayEqualityComparer();
+        private static readonly StringArrayEqualityComparer StringArrayComparer = new StringArrayEqualityComparer();
         private readonly object lockZeroTags = new object();
         private readonly HashSet<string> tagKeysInteresting;
         private readonly int tagsKeysInterestingCount;
 
+        private readonly ConcurrentDictionary<string[], string[]> tagKeyCombinations =
+            new ConcurrentDictionary<string[], string[]>(StringArrayComparer);
+
         // Two-Level lookup. TagKeys x [ TagValues x Metrics ]
         private readonly ConcurrentDictionary<string[], ConcurrentDictionary<object[], int>> keyValue2MetricAggs =
-            new ConcurrentDictionary<string[], ConcurrentDictionary<object[], int>>(new StringArrayEqualityComparer());
+            new ConcurrentDictionary<string[], ConcurrentDictionary<object[], int>>(StringArrayComparer);
```

Comment:
another alternate option which doesn't have the risk of too many entries (when user keeps providing keys in different order)

Have the dictionary as before
If tagKeys lookup fail, sort and lookup again.
If fails, insert both original tagKeys and its sorted one to the dictionary.

So that we only store atmost 2 entries per key set. And we only do a single lookup in hotpath, as opposed to 2 look ups.

Reply:
That's a good suggestion. The only issue with this would be if the user provides the sorted combination as the very first combination and uses some random combination later on. In this case, we would always be sorting the keys.

Reply:
yes, if the same order is re-used, then you get max performance. else lower perf.

Reply:
I agree this is good optimization to try out. Reusing the same order is probably the most common scenario. I suppose it's possible for library to use a single instrument in multiple code paths which add dimensions in a different order, but probably an edge case.

A different order may be likely in the event I have two libraries emitting the same metric name, but since they're different libraries they'd be separate Metric instances, right?

Reply:
Based on offline sync, this may not be feasible. 
We can come back to this and keep optimizing. For now, this PR avoids sorting in hot path, and makes a very significant perf boost.

Reply:
> A different order may be likely in the event I have two libraries emitting the same metric name, but since they're different libraries they'd be separate Metric instances, right?

They cannot emit with same metric name (unless different Meter). So it'll be different instances, yes.

Reply:
> Based on offline sync, this may not be feasible.

I was wondering about this myself. Was the issue that synchronizing the two inserts (sorted and original order) be tough?

Reply:
The problem here was keeping the two entries for a given combination in sync.

For example:

`counter.Add(10, new("A",1), new("C",3), new("B",2));` 

Here, since, ("A","C","B") is not present in the dictionary, we add these two entries:

("A","C","B") -> (1,3,2) -> _MetricPointIndex1_

("A","B","C") -> (1,2,3) -> _MetricPointIndex1_

Now, if we encounter a

`counter.Add(10, new("A",10), new("B",20), new("C",30));` 

we will add another entry to the inner dictionary:

<pre>
("A","B","C") -> (1,2,3) -> _MetricPointIndex1_
              -> (10,20,30)-> _MetricPointIndex2_
</pre>

This is fine but we also need to add this entry to the below:

<pre>
("A","C","B") -> (1,3,2) -> _MetricPointIndex1_
             <b> -> (10,30,20)-> _MetricPointIndex2_ </b> (This is the difficult part)
</pre>

When we get a new set of tag values we have to find if there is another tag key combination present in the dictionary, and if it's present we have to add the same `MetricPointIndex` for the tag values sorted according to the tag keys.
In this case, we get the new tag values (10,20,30) which are sorted by ("A","B","C"). Now we have to find if there is some other combination of ("A","B","C") present in the dictionary. If it's present we then have to sort the tag values according to the combination that is present, in this case, ("A","C","B") which would mean we have to sort the tag values like this: (10,30,20). We then have to ensure that we assign it the same `MetricPoint` index.

Reply:
Ugh... right gotta deal with the values as well.